numpy/_core/src/umath/string_buffer.h

```diff
@@ -0,0 +1,279 @@
+{
+    *bytes = 4;
+    return *(npy_ucs4 *)buf;
+}
+
+
+template <>
+inline npy_ucs4
+getchar<ENCODING::UTF8>(const unsigned char *buf, int *bytes)
+{
+    if (buf[0] <= 0x7F) {
+        // 0zzzzzzz -> 0zzzzzzz
+        *bytes = 1;
+        return (npy_ucs4)(buf[0]);
+    }
+    else if (buf[0] <= 0xDF) {
+        // 110yyyyy 10zzzzzz -> 00000yyy yyzzzzzz
+        *bytes = 2;
+        return (npy_ucs4)(((buf[0] << 6) + buf[1]) - ((0xC0 << 6) + 0x80));
+    }
+    else if (buf[0] <= 0xEF) {
+        // 1110xxxx 10yyyyyy 10zzzzzz -> xxxxyyyy yyzzzzzz
+        *bytes = 3;
+        return (npy_ucs4)(((buf[0] << 12) + (buf[1] << 6) + buf[2]) -
+                          ((0xE0 << 12) + (0x80 << 6) + 0x80));
+    }
+    else {
+        // 11110www 10xxxxxx 10yyyyyy 10zzzzzz -> 000wwwxx xxxxyyyy yyzzzzzz
+        *bytes = 4;
+        return (npy_ucs4)(((buf[0] << 18) + (buf[1] << 12) + (buf[2] << 6) + buf[3]) -
+                          ((0xF0 << 18) + (0x80 << 12) + (0x80 << 6) + 0x80));
+    }
+}
+
+
+template <ENCODING enc>
+struct Buffer {
+    char *buf;
+    char *after;
+
+    inline Buffer<enc>() {
+        buf = after = NULL;
+    }
+
+    inline Buffer<enc>(char *buf_, int elsize_) {
+        buf = buf_;
+        after = buf_ + elsize_;
+    }
+
+    inline npy_int64 length() {
+        Buffer tmp(after, 0);
+        tmp--;
+        while (tmp >= *this && *tmp == '\0') {
+            tmp--;
+        }
+        return (npy_int64) (tmp - *this + 1);
+    }
+
+    inline Buffer<enc>& operator+=(npy_int64 rhs) {
+        switch (enc) {
+        case ENCODING::ASCII:
+            buf += rhs;
+            break;
+        case ENCODING::UTF8:
+            int bytes;
+            getchar<enc>((const unsigned char *) buf, &bytes);
+            buf += rhs * bytes;
+            break;
+        case ENCODING::UTF32:
+            buf += rhs * sizeof(npy_ucs4);
+            break;
+        }
+        return *this;
+    }
+
+    inline Buffer<enc>& operator-=(npy_int64 rhs) {
+        switch (enc) {
+        case ENCODING::ASCII:
+            buf -= rhs;
+            break;
+        case ENCODING::UTF8: {
+            for (int i = 0; i < rhs; i++) {
+                for (int j = 0; j < 4; j++) {
+                    buf--;
+                    if (MSB(*buf) == 0) {
+                        break;
+                    }
+                }
+            }
+        }
+        case ENCODING::UTF32:
+            buf -= rhs * sizeof(npy_ucs4);
+            break;
+        }
+        return *this;
+    }
+
+    inline Buffer<enc>& operator++() {
+        *this += 1;
+        return *this;
+    }
+
+    inline Buffer<enc> operator++(int) {
+        Buffer<enc> old = *this;
+        operator++();
+        return old; 
+    }
+
+    inline Buffer<enc>& operator--() {
+        *this -= 1;
+        return *this;
+    }
+
+    inline Buffer<enc> operator--(int) {
+        Buffer<enc> old = *this;
+        operator--();
+        return old; 
+    }
```

Comment:
My feeling is that it's okay, since it's being used the same way a buffer would be used. The fact it doesn't hold a buffer internally seems like an implementation detail in the eyes of the loops that'll be using it.